rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,344 @@
+events are not actually delivered when they are buffered, the final
+delivery status will have to be confirmed after the event is loaded back
+from the buffer. The source, to be used to determine where
+to provide the finalization status indicator, must be converted into a form
+that may be serialized (ie string or integer index). However, the
+configuration may be reloaded while the event is buffered in such a way
+that the source name provided in the configuration may change, or a
+different component may be substituted for the same identifier. As such,
+the configured component name is not sufficient to uniquely identify the
+source. Further, Vector may be stopped while the event is buffered and
+then restarted with an identical configuration. At this point, the old
+source will be distinct from the current source, despite having the same
+configured name. As such, the persisted identifier for the source must
+stay the same across configuration reloads, but must be different for
+each run of Vector.
+
+## Internal Proposal
+
+There are two components to the proposal: the communication between
+event finalization and the sources, and the resulting event
+metadata required to track the event status.
+
+### Event finalization
+
+When instantiated, the topology will provide each source with a unique
+identifier for the token. The source may use this identifier to create
+one or more notification channels, each with their own unique
+identifier. This provides both a unique identifier of the source that
+needs to receive the event finalization status as well as the mechanism
+for delivering the event identifier and status.
+
+The finalization status is recorded when the last copy of the event is
+dropped. At that time, the finalization status stored in the event is
+recorded in a batch status buffer. Since events may be internally
+dropped when they are merged, a special "no-op" status is used to
+indicate that no status needs to be recorded to prevent superfluous
+status updates. When the last copy of the batch status is dropped, it is
+sent to the source.
+
+### Event metadata
+
+Given the channel above, the following source metadata will be added to events:
+
+1. A finalization status, to be delivered to all sources.
+2. An array of event sources.
+
+Each event source is a shared reference to a batch notifier, containing:
+
+1. The final status indicator. This will default to a success indicator,
+   but may be set to indicate failure if any event in the batch fails
+   delivery.
+2. A one-shot notification channel to the source.
+3. The unique identifier used to recreate the channel after deserialization.
+
+### Data structures
+
+```rust
+struct EventMetadata {
+    // … existing fields …
+    finalizer: Option<Arc<EventFinalizer>>,
+}
+
+struct EventFinalizer {
+    status: EventStatus,
+    sources: Box<[Arc<BatchNotifier>]>,
+}
+
+struct BatchNotifier {
+    status: Mutex<BatchStatus>,
```

Comment:
When the event is dropped, all source batches are updated from the status stored for that event. This may happen from multiple threads, hence the `Mutex`. An atomic should work here too, but this makes it clear what kind of data it contains since there isn't a generic `Atomic<T>` type.

The data flow for notifications is:
1. Each sink will update the event's finalizer `status`.
2. If one of those sinks is configured to be authoritative, it will immediately update the status of all its source batches and mark its own status as `NoOp`.
3. Otherwise when the last copy of the event is dropped, it does the above status update.
4. When the last event of a batch is dropped, the batch notifier sends a single trigger (hence oneshot) to the source, which can then ack the batch.

The point here being to minimize the number of notifications. If we determine we want to selectively ack the batch, that final oneshot will hand back an array instead of a single value, but that is not needed up front AFAIK.

I'll add clarification on this to the RFC.

Reply:
Apparently `oneshot::Sender` is a single word, probably a heap pointer.

Reply:
Thanks for the clarification. Makes good sense to me.